### PR TITLE
DATAES-220 - First version of Jest Implementation

### PR DESCRIPTION
**Pull request by VanRoy**
With the support of @pulse00 we have polished the first implementation of Jest as ElasticSearchTemplate backend.

See https://jira.spring.io/browse/DATAES-220 for more info.

I have signed and agree to the terms of the Spring Individual Contributor License Agreement.

### CODE REVIEW

**Comment by cforce** [comment]
After merge i had to add 
org.apache.commons.lang3.*;version="${commonslang:[=.=.=,+3.3.2)}",

into manifest so it compiled

**Comment by akonczak** [comment]
Hi @VanRoy , @pulse00

Great job with the integration jest with spring data.

Last week we had a discussion with elasticsearch people and its seem that in version 5.x they might have low level Rest client support.

More information can be found here: 
https://github.com/elastic/elasticsearch/pull/18735

In this circumstance we will wait for the next release of ES before we will make any decision around supporting this feature. 

Thanks, 
Artur

**Comment by VanRoy** [comment]
Hi @akonczak,

I understand your decision about that. On my projet we waiting for long time ago the native support of REST client too.

Do you see any inconveniences to I package my JestElasticSearchTemplate in new and independent Github projet and in the Maven Central ?

Thank a lot.
Julien.

**Comment by akonczak** [comment]
Hi @olivergierke 

Would you be able to help us with this question Oliver ?

Thanks, 
Artur

**Comment by VanRoy** [comment]
Hi @olivergierke , @akonczak ,

You can see the projet in this Github repository : https://github.com/VanRoy/spring-data-jest

I will waiting for your agreement to publish to maven central.

Thanks.
Julien.

**Comment by VanRoy** [comment]
Hi @akonczak , @olivergierke ,

Finally, I have published the artefact on maven central.

Do not hesitate to come back to me if any problems related to this publicated.

Thank a lot.
Julien.

**Comment by cforce** [comment]
I read a notice on spring blog that there i no support for Jest . Does it mean this has been merged to a release ?  Which one , because the most current on spring repo is still the old spring-data-elasticsearch 1.1.0


https://spring.io/blog/2016/07/12/this-week-in-spring-july-12th-2016

> The first release candidate of Spring Boot 1.4 is now available, packed with a lot of great things! This new release features, among other things, a unified @EntityScan for JPA, MongoDB, Neo4j, Couchbase and Cassandra, an auto-configured RestTemplateBuilder, support for pure REST client tests via @RestClientTest, support for Jest (the Elasticsearch REST client), and upgrades to Spring Integration 4.3, Spring AMQP 1.6, Spring REST Docs 1.1, MongoDB Java Driver 3 and more! Download the bits and kick the tires right now!

**Comment by jpblair** [comment]
I would love to be able to use this with an AWS Elasticsearch cluster.  Any estimate when this might be merged in/see a release?  Also, will it only be compatible with 2.x versions, or 1.x as well?

**Comment by pulse00** [comment]
> In this circumstance we will wait for the next release of ES before we will make any decision around supporting this feature.

@akonczak looks like the low level rest client has been merged into elasticsearch. Is there any place i can watch (jira / github issue) to be up to date on what's happening with regards to spring-data-elasticsearch?

**Comment by akonczak** [comment]
@pulse00 I have created a ticket to look for new ES release and do some prep work, on of the point is to provide a support for native rest ES client.

**Comment by akonczak** [comment]
@cforce the announcement is about a spring-boot not about spring-data-elasticsearch.

File: src/main/java/org/springframework/data/elasticsearch/core/JestElasticsearchTemplate.java [hunk 1]
```diff
@@ -0,0 +1,1283 @@
...
+					indexBuilder.index(indexName).type(type);
+				}
+			} else if (query.getSource() != null) {
+				indexBuilder = new Index.Builder(query.getSource()).index(indexName).type(type).id(query.getId());
+			} else {
+				throw new ElasticsearchException("object or source is null, failed to index the document [id: " + query.getId() + "]");
+			}
+			if (query.getVersion() != null) {
+				indexBuilder.setParameter(Parameters.VERSION, query.getVersion());
+				indexBuilder.setParameter(Parameters.VERSION_TYPE, EXTERNAL);
```
Comment (ausov) [inline src/main/java/org/springframework/data/elasticsearch/core/JestElasticsearchTemplate.java]: Elasticsearch expects `version_type` parameter value to be lowercase (case-sensitive), i.e. it should be "external", not "EXTERNAL" . Suggestion: use `EXTERNAL.name().toLowerCase()`.

**Comment by VanRoy** [comment]
@ausov The branch will certainly never be merged.

Try rather my project pending the final SpringDataES solution ( https://github.com/VanRoy/spring-data-jest )

**Comment by ausov** [comment]
@VanRoy ok, thank you.

**Comment by sothawo** [comment]
Sorry, a very late answer; in the meantime REST support is available in version (3.2.0.RC1). So there is no need for this patch. Thanks for your work anyway.

**Pull request closed without merging**